File: python_pipeline/validate_core.py
#!/usr/bin/env python3
"""Numerical validation of the Python port against reference values verified
during the code review (which matched the magnitudes in the MATLAB draft).

Run from py_port:  python validate_core.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np

from gar import build_datasets, qr_direct, fit_skewt_quantiles, skewt_ppf

ROOT = Path(__file__).resolve().parent
FILES = {
    "aggregate": ROOT / "input" / "DataIP_GaR.csv",
    "brent": ROOT / "input" / "DataOIL.csv",
    "panel": ROOT / "input" / "country_sector.csv",
    "eu_sectors": {},
}
TAUS = np.round(np.arange(0.05, 0.951, 0.05), 4)

results = []


def check(name, value, ref, tol):
    ok = np.isfinite(value) and abs(value - ref) <= tol
    print(f"{name:<55s} py = {value:+.4f} | ref = {ref:+.4f} | "
          f"{'PASS' if ok else 'FAIL'}")
    results.append(ok)


def main():
    np.random.seed(123)
    dl = build_datasets(FILES)
    d = dl["aggregate"]

    # 1. unconditional correlations
    ok = d["Diff_IP"].notna() & d["Diff_GP"].notna() & d["Diff_BP"].notna()
    check("corr(Diff_IP, Diff_GP), 2004-2023",
          d.loc[ok, "Diff_IP"].corr(d.loc[ok, "Diff_GP"]), 0.077, 0.02)
    check("corr(Diff_IP, Diff_BP), 2004-2023",
          d.loc[ok, "Diff_IP"].corr(d.loc[ok, "Diff_BP"]), 0.499, 0.02)

    # 2. NGPI construction
    check("max(NGPI) aggregate (100*log scale)",
          float(d["NGPI"].max()), 37.49, 0.5)

    # 3. aggregate Q05 NGPI coefficients
    okn = d["Diff_IP"].notna() & d["NGPI"].notna()
    y = d.loc[okn, "Diff_IP"].to_numpy()
    x = d.loc[okn, "NGPI"].to_numpy()
    res = qr_direct(y, x, horizons=[1, 12], taus=TAUS, include_lag_y=True,
                    target="point", var_names=["Intercept", "NGPI", "lag"])
    j05 = int(np.argmin(np.abs(TAUS - 0.05)))
    check("Aggregate Q05 NGPI coefficient, h = 1",
          res[1]["coef"].loc["NGPI"].iloc[j05], -0.073, 0.03)
    check("Aggregate Q05 NGPI coefficient, h = 12",
          res[12]["coef"].loc["NGPI"].iloc[j05], -0.196, 0.04)

    # 4. Germany energy sector Q95 NGPI at h = 12
    gkey = next((k for k in dl if k.startswith("Germany.")
                 and "energy" in k.lower()), None)
    if gkey:
        g = dl[gkey]
        okg = g["Diff_IP"].notna() & g["NGPI"].notna()
        resg = qr_direct(g.loc[okg, "Diff_IP"].to_numpy(),
                         g.loc[okg, "NGPI"].to_numpy(),
                         horizons=[12], taus=TAUS, include_lag_y=True,
                         target="point",
                         var_names=["Intercept", "NGPI", "lag"])
        j95 = int(np.argmin(np.abs(TAUS - 0.95)))
        check("Germany energy Q95 NGPI coefficient, h = 12",
              resg[12]["coef"].loc["NGPI"].iloc[j95], -0.145, 0.04)
    else:
        print("Germany energy dataset not found -- skipped check 4")

    # 5. averaged target equals point target at h = 1
    res_avg = qr_direct(y, x, horizons=[1], taus=TAUS, include_lag_y=True,
                        target="average", var_names=["Intercept", "NGPI", "lag"])
    check("Averaged == point target at h = 1 (Q05 NGPI)",
          res_avg[1]["coef"].loc["NGPI"].iloc[j05],
          res[1]["coef"].loc["NGPI"].iloc[j05], 1e-8)

    # 6. skew-t round trip: fit on known quantiles, recover them
    true = dict(xi=-0.5, omega=2.0, alpha=-3.0, nu=6.0)
    q_true = skewt_ppf(TAUS, **true)
    ft = fit_skewt_quantiles(q_true, TAUS)
    q_fit = skewt_ppf(TAUS, ft["xi"], ft["omega"], ft["alpha"], ft["nu"])
    check("Skew-t round trip: max |q_fit - q_true|",
          float(np.max(np.abs(q_fit - q_true))), 0.0, 0.05)

    print(f"\n{sum(results)} of {len(results)} checks passed.")
    if not all(results):
        raise SystemExit("Validation failed -- inspect the FAIL lines above.")


if __name__ == "__main__":
    main()

File: r_pipeline/validate_core.R
# validate_core.R -- quick numerical validation of the R port against
# independently verified reference values (computed from the same CSVs with
# Python/statsmodels during the code review; those in turn matched the
# magnitudes reported in the MATLAB-based draft).
#
# Run AFTER 00_read_data.R:   Rscript validate_core.R
# Every check prints the R value, the reference, and PASS/FAIL (tolerances
# are loose for the bootstrap-free point estimates; QR solvers may differ
# in the last decimals when the solution is non-unique).

cfg <- list(root = normalizePath("."),
            files = list(
              aggregate = "input/DataIP_GaR.csv",
              brent     = "input/DataOIL.csv",
              panel     = "input/country_sector.csv",
              eu_sectors = list()),
            horizons = c(1, 12),
            taus = seq(0.05, 0.95, by = 0.05),
            n_boot = 0, use_cache = FALSE, seed = 123)

set.seed(cfg$seed)
source(file.path(cfg$root, "scripts", "00_read_data.R"))
source(file.path(cfg$root, "R", "quantile_reg.R"))

dl <- readRDS(file.path(cfg$root, "data", "prepared.rds"))

check <- function(name, value, ref, tol) {
  ok <- is.finite(value) && abs(value - ref) <= tol
  cat(sprintf("%-55s R = %+.4f | ref = %+.4f | %s\n",
              name, value, ref, if (ok) "PASS" else "FAIL"))
  ok
}

results <- c()

## 1. Unconditional correlations --------------------------------------------
d <- dl$aggregate
ok <- stats::complete.cases(d$Diff_IP, d$Diff_GP, d$Diff_BP)
results <- c(results,
  check("corr(Diff_IP, Diff_GP), 2004-2023",
        stats::cor(d$Diff_IP[ok], d$Diff_GP[ok]), 0.077, 0.02),
  check("corr(Diff_IP, Diff_BP), 2004-2023",
        stats::cor(d$Diff_IP[ok], d$Diff_BP[ok]), 0.499, 0.02))

## 2. NGPI construction -------------------------------------------------------
results <- c(results,
  check("max(NGPI) aggregate (100*log scale)",
        max(d$NGPI, na.rm = TRUE), 37.49, 0.5))

## 3. Aggregate Q5 NGPI coefficient at h = 1 and h = 12 ----------------------
okn <- stats::complete.cases(d$Diff_IP, d$NGPI)
y <- d$Diff_IP[okn]; x <- d$NGPI[okn]
res <- qr_direct(y, matrix(x, dimnames = list(NULL, "NGPI")),
                 horizons = c(1, 12), taus = cfg$taus,
                 include_lag_y = TRUE, target = "point",
                 var_names = c("Intercept", "NGPI", "Diff_IP_lag"))
b_h1  <- res[["1"]]$coef["NGPI", which.min(abs(cfg$taus - 0.05))]
b_h12 <- res[["12"]]$coef["NGPI", which.min(abs(cfg$taus - 0.05))]
results <- c(results,
  check("Aggregate Q05 NGPI coefficient, h = 1",  b_h1,  -0.073, 0.03),
  check("Aggregate Q05 NGPI coefficient, h = 12", b_h12, -0.196, 0.04))

## 4. Germany energy sector, Q95 NGPI at h = 12 ------------------------------
gkey <- grep("^Germany\\..*energy", names(dl), ignore.case = TRUE, value = TRUE)
if (length(gkey) == 1) {
  g <- dl[[gkey]]
  okg <- stats::complete.cases(g$Diff_IP, g$NGPI)
  resg <- qr_direct(g$Diff_IP[okg],
                    matrix(g$NGPI[okg], dimnames = list(NULL, "NGPI")),
                    horizons = 12, taus = cfg$taus,
                    include_lag_y = TRUE, target = "point",
                    var_names = c("Intercept", "NGPI", "Diff_IP_lag"))
  bg <- resg[["12"]]$coef["NGPI", which.min(abs(cfg$taus - 0.95))]
  results <- c(results,
    check("Germany energy Q95 NGPI coefficient, h = 12", bg, -0.145, 0.04))
} else {
  cat("Germany energy dataset not found -- skipped check 4\n")
}

## 5. Averaged-target sanity: h = 1 equals point target ----------------------
res_avg <- qr_direct(y, matrix(x, dimnames = list(NULL, "NGPI")),
                     horizons = 1, taus = cfg$taus,
                     include_lag_y = TRUE, target = "average",
                     var_names = c("Intercept", "NGPI", "Diff_IP_lag"))
results <- c(results,
  check("Averaged target == point target at h = 1 (Q05 NGPI)",
        res_avg[["1"]]$coef["NGPI", which.min(abs(cfg$taus - 0.05))], b_h1, 1e-6))

cat(sprintf("\n%d of %d checks passed.\n", sum(results), length(results)))
if (!all(results)) stop("Validation failed -- inspect the FAIL lines above.")
